### PR TITLE
Bug 1394939 - XCUITrackingProtection testcase

### PR DESCRIPTION
**Pull request by ugarlapati**

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
TrackingProtectionTests.swift | 1 | Files should have consistent header comments. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Client.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -4391,7 +4394,7 @@
 					};
 					3BFE4B061D342FB800DDF53F = {
 						CreatedOnToolsVersion = 7.3.1;
-						DevelopmentTeam = 43AQ936H96;
+						DevelopmentTeam = 9G8J6YA743;
```
Comment (garvankeeley) [inline Client.xcodeproj/project.pbxproj]: Unstage

File: XCUITests/TrackingProtectionTests.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    override func tearDown() {
+        navigator = nil
+        app = nil
+        super.tearDown()
+    }
+    
+    func testTrackingProtection() {
+        navigator.goto(SettingsScreen)
+        let appsettingstableviewcontrollerTableviewTable = app.tables["AppSettingsTableViewController.tableView"]
+        while app.staticTexts["Tracking Protection"].exists == false {
```
Comment (garvankeeley) [inline XCUITests/TrackingProtectionTests.swift]: nit: `== false` to `!app`

Comment (garvankeeley) [reply XCUITests/TrackingProtectionTests.swift]: BTW, this block is cool, I didn't know an easy way to get XCUITest to see table view rows that were offscreen

File: XCUITests/TrackingProtectionTests.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    override func tearDown() {
+        navigator = nil
+        app = nil
+        super.tearDown()
+    }
+    
+    func testTrackingProtection() {
+        navigator.goto(SettingsScreen)
+        let appsettingstableviewcontrollerTableviewTable = app.tables["AppSettingsTableViewController.tableView"]
```
Comment (garvankeeley) [inline XCUITests/TrackingProtectionTests.swift]: nit: `let appSettingsTableView =`

File: XCUITests/TrackingProtectionTests.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        XCTAssertTrue(app.tables.cells["Private Browsing Mode Only"].isSelected)
+        XCTAssertFalse(app.tables.cells["Never"].isSelected)
+        
+        app.tables.cells["Always On"].tap()
+        XCTAssertTrue(app.tables.cells["Always On"].isSelected)
+        
+        app.navigationBars["Tracking Protection"].buttons["Settings"].tap()
+        
+        waitforExistence(app.navigationBars["Settings"].buttons["Done"])
+        app.navigationBars["Settings"]/*@START_MENU_TOKEN@*/.buttons["Done"]/*[[".buttons[\"Done\"]",".buttons[\"AppSettingsTableViewController.navigationItem.leftBarButtonItem\"]"],[[[-1,1],[-1,0]]],[1]]@END_MENU_TOKEN@*/.tap()
```
Comment (garvankeeley) [inline XCUITests/TrackingProtectionTests.swift]: Are these comments needed?

**Comment by garvankeeley** [comment]
@npark-mozilla excuse the drive-by review, I got excited to see this :), I am not 100% confident with my XCUITest skillz so please double-check my review as these are the trickiest tests IMO.

File: XCUITests/TrackingProtectionTests.swift (hunk 4 above)
Comment (npark-mozilla) [reply XCUITests/TrackingProtectionTests.swift]: ah, they are autogenerated by Xcode I believe

File: XCUITests/TrackingProtectionTests.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        navigator = createScreenGraph(app).navigator(self)
+    }
+
+    override func tearDown() {
+        navigator = nil
+        app = nil
+        super.tearDown()
+    }
+    
+    func testTrackingProtection() {
```
Comment (npark-mozilla) [inline XCUITests/TrackingProtectionTests.swift]: add comment to describe what the test does

File: XCUITests/TrackingProtectionTests.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        navigator.goto(SettingsScreen)
+        let appsettingstableviewcontrollerTableviewTable = app.tables["AppSettingsTableViewController.tableView"]
+        while app.staticTexts["Tracking Protection"].exists == false {
+            appsettingstableviewcontrollerTableviewTable.swipeUp()
+        }
+        appsettingstableviewcontrollerTableviewTable.staticTexts["Tracking Protection"].tap()
+        
+        XCTAssertFalse(app.tables.cells["Always On"].isSelected)
```
Comment (npark-mozilla) [inline XCUITests/TrackingProtectionTests.swift]: comment on why are we checking this

File: XCUITests/TrackingProtectionTests.swift [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        
+        let appsettingstableviewcontrollerTableviewTable1 = app.tables["AppSettingsTableViewController.tableView"]
+        while app.staticTexts["Tracking Protection"].exists == false {
+            appsettingstableviewcontrollerTableviewTable1.swipeUp()
+        }
+        
+        appsettingstableviewcontrollerTableviewTable1.staticTexts["Tracking Protection"].tap()
+        waitforExistence(app.tables.cells["Always On"])
+        
+        XCTAssertTrue(app.tables.cells["Always On"].isSelected)
```
Comment (npark-mozilla) [inline XCUITests/TrackingProtectionTests.swift]: comment why are we checking this

**Comment by garvankeeley** [comment]
@npark-mozilla you pinged me about testing the rules are applied. Just mentioning I am putting together some methods, just deciding on the easiest way.

**Comment by garvankeeley** [comment]
We could do something like this: http://garvankeeley.github.io/web/tracking-protection-test.html

This loads a stable blocking rule (both because the rule is unlikely to change, and the resource we load -the favicon- is likely to always be available).
The source for this page does:
```
 window.onload = () => {
            testImage('http://ymail.com/favicon.ico');
 }
```
And testImage function will show an alert if the image doesn't load. I assumed an alert is an easy to thing to check for in the UITest, if not, we can have some other action happen.

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
TabTrayController.swift | 471 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
TabTrayController.swift | 1154 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by npark-mozilla** [comment]
this is now moved to https://github.com/mozilla-mobile/firefox-ios/pull/3141

**Pull request merged**